### PR TITLE
try to fix sub_command list index out of range

### LINKED ISSUE #589: ERROR: Rosdep experienced an error: list index out of range in debian stretch (installation from source)
```
rosdep version: 0.12.1

Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/rosdep2/main.py", line 140, in rosdep_main
    exit_code = _rosdep_main(args)
  File "/usr/lib/python2.7/dist-packages/rosdep2/main.py", line 388, in _rosdep_main
    return _package_args_handler(command, parser, options, args)
  File "/usr/lib/python2.7/dist-packages/rosdep2/main.py", line 484, in _package_args_handler
    return command_handlers[command](lookup, packages, options)
  File "/usr/lib/python2.7/dist-packages/rosdep2/main.py", line 706, in command_install
    installer.install(uninstalled, **install_options)
  File "/usr/lib/python2.7/dist-packages/rosdep2/installers.py", line 525, in install
    verbose=verbose, quiet=quiet)
  File "/usr/lib/python2.7/dist-packages/rosdep2/installers.py", line 588, in install_resolved
    if isinstance(sub_command[0], list):  # list of alternatives
IndexError: list index out of range
```

while installing ros on my beagleboneblack(armhf) with debian stretch, from source this issue was found

### PR DESCRIPTION
**Pull request by flixr**
should fix #589

### CODE REVIEW

**Comment by wjwwood** [comment]
@ipa-mdl can you review this please (asap since it appears to be affecting people in the wild). I would like to know what the root cause is. This pr will certainly avoid the crash, but I'm wondering why `sub_command` is `None` in the first place.

**Comment by flixr** [comment]
While I haven't done much debugging, I actually don't think it's `None`... just empty

**Comment by wjwwood** [comment]
Yeah, sorry, you're right, got it mixed up with another issue (different repository), but still it'd be nice to know if the empty list is intentional or not.

File: src/rosdep2/installers.py [hunk 1]
```diff
@@ -561,9 +561,11 @@ def install_resolved(self, installer_key, resolved, simulate=False, interactive=
         if simulate:
             print('#[%s] Installation commands:' % (installer_key))
             for sub_command in command:
+                if not sub_command:
+                    continue
                 if isinstance(sub_command[0], list):
                     sub_cmd_len = len(sub_command)
-                    for cmd, i in enumerate(sub_command):
+                    for i, cmd in enumerate(sub_command):
```
Comment (ipa-mdl) [inline src/rosdep2/installers.py]: :+1: 
Not sure how this has slipped through..

**Comment by ipa-mdl** [comment]
I cannot find any reason why `sub_command` should be empty (or `None`). There might be an underlying issue that would get hidden by this fix.

**Comment by flixr** [comment]
I think if there is nothing to install, it is empty...

**Comment by flixr** [comment]
I think the top level list still contains one empty list...

**Comment by ipa-mdl** [comment]
I have added some fixes in #585, at least the simulation issue would be fixed.
It might even fix this problem with empty lists etc.
@flixr: please try it.

**Comment by flixr** [comment]
With #585 I get:
```
$ rosdep install --from-paths . --ignore-src -r -y --simulate                                                                            

ERROR: Rosdep experienced an internal error.
Please go to the rosdep page [1] and file a bug report with the message below.
[1] : http://www.ros.org/wiki/rosdep

rosdep version: 0.12.1

Bad installer [apt]: 'account-plugin-aim'
```

**Comment by ipa-mdl** [comment]
@flixr: Thanks!
Are you using any special rosdep setup?

**Comment by ipa-mdl** [comment]
Something related to `account-plugin-aim`?
Please try to call it with `ROSDEP_DEBUG=1` set as enviroment variable to see the stacktrace.

**Comment by flixr** [comment]
> Something related to account-plugin-aim?

I'm pretty sure it has nothing to do with `account-plugin-aim`, the original error also happens on our CI using the `ros:kinetic-perception` docker image...

> Please try to call it with ROSDEP_DEBUG=1 set as enviroment variable to see the stacktrace.

```
ROSDEP_DEBUG=1 rosdep install --from-paths . --ignore-src -r -y --simulate                                                             
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/ruby.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/ruby.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/indigo/distribution.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/indigo/distribution.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/ruby.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/ruby.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/indigo/distribution.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/indigo/distribution.yaml]: []
_load_view_dependencies[*default*]
_load_view_dependencies[*default*]: ['sources.list']
_load_view_dependencies[sources.list]
_load_view_dependencies[*default*]
_load_view_dependencies[*default*]: ['sources.list']
_load_view_dependencies[sources.list]
Traceback (most recent call last):
  File "/home/ruess/.local/lib/python2.7/site-packages/rosdep2/installers.py", line 459, in get_uninstalled
    packages_to_install = installer.get_packages_to_install(resolved)
  File "/home/ruess/.local/lib/python2.7/site-packages/rosdep2/installers.py", line 375, in get_packages_to_install
    detected = self.detect_fn(resolved)
  File "/home/ruess/.local/lib/python2.7/site-packages/rosdep2/platforms/debian.py", line 184, in dpkg_detect
    virtual = [n for (n, v, pr) in remaining if v and len(dpkg_detect(pr)) > 0]
  File "/home/ruess/.local/lib/python2.7/site-packages/rosdep2/platforms/debian.py", line 179, in dpkg_detect
    installed_packages = [version_lock_map[r] for r in ret_list]
KeyError: 'account-plugin-aim'


ERROR: Rosdep experienced an internal error.
Please go to the rosdep page [1] and file a bug report with the message below.
[1] : http://www.ros.org/wiki/rosdep

rosdep version: 0.12.1

Bad installer [apt]: 'account-plugin-aim'
```

**Comment by ipa-mdl** [comment]
This really looks like a hidden issuem which would get covered by this PR.

I have just noticed that e.g. `apt-cache show curl` returns two version sets.
I guess that `_read_apt_cache_showpkg` does not parse the output properly.
Given the fact that your server keeps multiple versions as well, this might be the underlying problem.

**Comment by flixr** [comment]
> This really looks like a hidden issuem which would get covered by this PR.

Maybe... but it worked fine with earlier rosdep versions...

**Comment by ipa-mdl** [comment]
> Maybe... but it worked fine with earlier rosdep versions...

There seems to be a regression, that's for sure.
Please try to add more outputs and check them for suspicous lines.
It's really hard to tell without an example workspace..

**Comment by flixr** [comment]
So `_read_apt_cache_showpkg` yields an empty list as third element in the tuple for packages that are not found.
In my case this is probably because the package is not yet released and hence not found with apt-cache.

If I fix the list of virtual packages with
```
virtual = [n for (n, v, pr) in remaining if v and pr and len(dpkg_detect(pr)) > 0]
```
I get the same `list index out of range` again, which is fixed by this PR.

**Comment by flixr** [comment]
You might also be able to reproduce this with a package that is already in the rosdep list but not yet available as deb (not yet synced) like e.g. `rc-genicam-api`

**Comment by ipa-mdl** [comment]
> So _read_apt_cache_showpkg yields an empty list as third element in the tuple for packages that are not found.

It should return `None`. I will add a unit test to #585.

> In my case this is probably because the package is not yet released and hence not found with apt-cache.

Packages that are not released (private or public) should not get resolved to the apt installer. (@wjwwood: correct?)

> You might also be able to reproduce this with a package that is already in the rosdep list but not yet available as deb (not yet synced) like e.g. rc-genicam-api

I would consider this a config error.
However, the code should print a proper error message instead of empty lists..

**Comment by ipa-mdl** [comment]
I have fixed another bug with regard to version locks.
@flixr: please try #585 again.

**Comment by flixr** [comment]
> Packages that are not released (private or public) should not get resolved to the apt installer. (@wjwwood: correct?)

I should have been more precise: the package is released (so found in rosdep) but still not available in apt-cache as it is still in shadow-fixed.

> I would consider this a config error.

I disagree. The package already needs this depend to properly build on the ROS buildfarm.
Just because it is not available in your local apt-cache (you actually just might not have done an apt update in a while), doesn't make this a config error.

Additionally the package in question (`rc_genicam_api`) is a third-party package that doesn't depend on ROS.
It's just released with ROS for convenience for other users and builds just fine without the "ROS packaged version" of that package.

> However, the code should print a proper error message instead of empty lists..

Info or warning maybe, but I wouldn't consider this an error.

**Comment by flixr** [comment]
> I have fixed another bug with regard to version locks.
> @flixr: please try #585 again.

Still the same since `_read_apt_cache_showpkg` still returns an empty list here:
https://github.com/ros-infrastructure/rosdep/blob/1e69f89988e8d2fbe931bc14639eb380bff30844/src/rosdep2/platforms/debian.py#L145
```
rosdep install --from-paths . --ignore-src -r -y --simulate
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/ruby.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/ruby.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/indigo/distribution.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/indigo/distribution.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/ruby.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/ruby.yaml]: []
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/indigo/distribution.yaml]
_load_view_dependencies[https://raw.githubusercontent.com/ros/rosdistro/master/indigo/distribution.yaml]: []
_load_view_dependencies[*default*]
_load_view_dependencies[*default*]: ['sources.list']
_load_view_dependencies[sources.list]
_load_view_dependencies[*default*]
_load_view_dependencies[*default*]: ['sources.list']
_load_view_dependencies[sources.list]
Traceback (most recent call last):
  File "/home/ruess/.local/lib/python2.7/site-packages/rosdep2/installers.py", line 459, in get_uninstalled
    packages_to_install = installer.get_packages_to_install(resolved)
  File "/home/ruess/.local/lib/python2.7/site-packages/rosdep2/installers.py", line 375, in get_packages_to_install
    detected = self.detect_fn(resolved)
  File "/home/ruess/.local/lib/python2.7/site-packages/rosdep2/platforms/debian.py", line 185, in dpkg_detect
    virtual = [n for (n, v, pr) in remaining if v and len(dpkg_detect(pr)) > 0]
  File "/home/ruess/.local/lib/python2.7/site-packages/rosdep2/platforms/debian.py", line 180, in dpkg_detect
    installed_packages = [version_lock_map[r] for r in ret_list]
KeyError: 'account-plugin-aim'


ERROR: Rosdep experienced an internal error.
Please go to the rosdep page [1] and file a bug report with the message below.
[1] : http://www.ros.org/wiki/rosdep

rosdep version: 0.12.1

Bad installer [apt]: 'account-plugin-aim'
```

**Comment by ipa-mdl** [comment]
> Still the same since _read_apt_cache_showpkg still returns an empty list here:

Could you provide me the output of the corresponding `apt-cache showpkg` invocation?
I guess there' something that gets parsed wrong.
Do you you virtual packages at all?

> > I would consider this a config error.

> I disagree. The package already needs this depend to properly build on the ROS buildfarm.
Just because it is not available in your local apt-cache (you actually just might not have done an apt update in a while), doesn't make this a config error.

If the package gets resolved by the Apt installer (and not as local ROS package), it should be available via apt. In my tests the system works for arbitrary package names, of course the install step fails afterwards.
The output parser seems to miss something here.
Currently, it considers all packages to be virtual if they don't have any version attached.
Perhaps this assumption is wrong for same cases.

**Comment by flixr** [comment]
> Could you provide me the output of the corresponding apt-cache showpkg invocation?

```
$ apt-cache showpkg ros-kinetic-rc-genicam-api
Package: ros-kinetic-rc-genicam-api
Versions: 

Reverse Depends: 
  ros-kinetic-rc-visard-driver,ros-kinetic-rc-genicam-api
Dependencies: 
Provides: 
Reverse Provides: 
```

So it seems that apt knows this package could exist since ros-kinetic-rc-visard-driver has an or-depends on it
```
Depends: libboost-system1.58.0, libc6 (>= 2.14), libconsole-bridge0.2v5, libgcc1 (>= 1:3.0), libprotobuf9v5, libstdc++6 (>= 5.2), rc-dynamics-api, rc-genicam-api, rc-genicam-api | ros-kinetic-rc-genicam-api, rc-dynamics-api | ros-kinetic-rc-dynamics-api
```

**Comment by flixr** [comment]
It seems to work with my `fix-virtual-packages` branch, which additionally has a commit that should fix  _read_apt_cache_showpkg: https://github.com/flixr/rosdep/commit/ec7602c67dedd5dcc5c6aff942a362f5b3c67136

@ipa-mdl you can cherry pick that if you want...

**Comment by ipa-mdl** [comment]
@flixr: Thanks, I will cherry-pick your solution and expand the unit tests.
Why did you change `+    starts.append(None)`? The `-1` was intended..

**Comment by flixr** [comment]
closing in favor of #585

**Pull request closed without merging**